[VIP-1715] Add real-time collaboration integration loader

shekharnwagh opened the pull request:
<!--
## For Automatticians!

:wave: Just a quick reminder that this is a public repo. Please don't include any internal links or sensitive data (like PII, private code, customer names, site URLs, etc. Any fixes related to security should be discussed with Platform before opening a PR. If you're not sure if something is safe to share, please just ask!

## For external contributors!

Welcome! We look forward to your contribution! ❤️
-->
## Description
<!--
A few sentences providing more context and describing the overall goals of the Pull Request.

Should include any special considerations, decisions, and links to relevant publicly available sources (e.g. GitHub issues)
-->

Implements `RealTimeCollaborationIntegration` class that loads custom build of `gutenberg` along with latest version of `vip-real-time-collaboration` plugin. Configures the `VIP_RTC_WS_AUTH_SECRET` and `VIP_RTC_WS_URL` constants from the `web_socket_auth_secret` and `web_socket_url` integration config values.

The custom build of gutenberg is built from https://github.com/Automattic/gutenberg/tree/add/experimental-collaborative-editing. It has the code required to support collaborative editing feature which is enabled by `vip-real-time-collaboration` plugin. This is temporary solution until the code from that branch makes it way to Gutenberg release.

It currently assumes only one version of the gutenberg custom build is available in `/vip-integrations/gutenberg` dir unlike the vip-real-time-collaboration plugin which can have multiple versions in `/vip-integrations/vip-real-time-collaboration-<version>/` dirs.

## Changelog Description

<!-- Changelogs are published for our customers. Well-written entries help them stay informed on platform changes and all of the great work that we do! -->

<!-- Write a concise description of changes in the relevant section.
- Add new line items as needed.
- Entries should follow the [Common Changelog Style Guide](https://github.com/vweevers/common-changelog). 
- Remove all unused sections before merging.
- Proof-read.
-->

### Added
- Add support for Real-Time Collaboration integration.

## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [ ] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test


1. Check out PR.
2. Add built code of gutenberg to `/vip-integrations/gutenberg` dir.
3. Add built code of `vip-real-time-collaboration` to `/vip-integrations/vip-real-time-collaboration-<version>/` dir.
4. Run `composer install` and `npm install` to install dependencies.
5. Activate the 'Real-Time Collaboration' integration on one of your test sites and note down its application ID.
6. Spin up a local dev for that site using command - `vip dev-env create @<application_id>.production -s <slug> -a <path_to_app_code> -u <path_to_local_vip-go-mu-plugins_dir>`
7. Start it using the command - `vip dev-env start --slug <slug> --editor vscode`
8. That should open a VS Code window. Modify the `<app_name>/integrations-config/integrations.json` to add following entry -

	```json
	"real-time-collaboration": {
		"type": "real-time-collaboration",
		"org": {
			"status": "enabled"
		},
		"env": {
			"status": "enabled",
			"config": {
				"web_socket_auth_secret": "test-secret-key",
				"web_socket_url": "wss://test.example.com/_ws"
			}
		}
	}
	```
9. Spin up the websocket server by running `npm run dev` in the `websocket-server` directory of `vip-real-time-collaboration` plugin local cloned directory. Make sure to set the `VIP_RTC_WS_AUTH_SECRET` env variable with same value as above.
10. Verify that the collaborative editing feature is working. 
11. Try modifying the `integrations.json` and see if the changes are reflected in the local dev.

shekharnwagh committed:
Add real-time collaboration integration

Implements RealTimeCollaborationIntegration class that loads custom
build of gutenberg along with latest version of vip-real-time-collaboration
plugin. Configures the VIP_RTC_WS_AUTH_SECRET and VIP_RTC_WS_URL
constants from the web_socket_auth_secret and web_socket_url integration
config values.

acicovic commented on tests/integrations/test-real-time-collaboration.php:
```diff
@@ -0,0 +1,154 @@
+		$integration_mock->expects( $this->never() )
+			->method( 'get_versions' );
+
+		$integration_mock->load();
```
We don't have an assertion here. Is this expected?

acicovic commented on tests/integrations/test-real-time-collaboration.php:
```diff
@@ -0,0 +1,154 @@
+			->getMock();
+
+		$integration_mock->method( 'get_env_config' )->willReturn( [
+			'web_socket_url' => 'wss://test.example.com/ws',
```
Unimportant, but if we want the `_` in `/ws`, let's add it.

shekharnwagh committed:
fix test to assert invocation after calling function

shekharnwagh committed:
align test example websocket url with real values

maxschmeling commented on vip-integrations.php:
```diff
@@ -34,6 +34,10 @@
 	require_once __DIR__ . '/integrations/jetpack.php';
 }
 
+if ( file_exists( __DIR__ . '/integrations/real-time-collaboration.php' ) ) {
```
Why do some integrations have an `if ( file_exists )` check?

shekharnwagh replied:
It came from @rinatkhaziev 's feedback on previous PR where we added RDB - https://github.com/Automattic/vip-go-mu-plugins/pull/6217#discussion_r2027614042

rinatkhaziev replied:
@maxschmeling because deploys are not-atomic and not properly gating this produces intermittent fatal errors.

maxschmeling commented on integrations/real-time-collaboration.php:
```diff
@@ -0,0 +1,111 @@
+		 */
+		$gutenberg_path = WPVIP_MU_PLUGIN_DIR . '/vip-integrations/gutenberg/gutenberg.php';
+		if ( file_exists( $gutenberg_path ) ) {
+			require_once $gutenberg_path;
```
Is this the right way to load the plugin over something like [`activate_plugin`](https://developer.wordpress.org/reference/functions/activate_plugin/)?

shekharnwagh replied:
I am not entirely sure, but all the integrations are activated this way. My guess (I might be wrong) is, we don't want to expose this as plugins in WP-Admin. @rinatkhaziev Do you the answer here ?

ingeniumed replied:
If this ensures that we don't expose Gutenberg in WP-Admin, I wonder if we might run into conflicts with those sites that have Gutenberg already enabled and want to use our plugin?

rinatkhaziev replied:
It would and this needs to be properly guarded against. Easy to test though :)

shekharnwagh replied:
Added logic to avoid loading the integration if Gutenberg is activated as a plugin - 7d1bf5fdd

ingeniumed commented on integrations/real-time-collaboration.php:
```diff
@@ -0,0 +1,111 @@
+		 * Get all the entries in the path of WPVIP_MU_PLUGIN_DIR/vip-integrations/vip-real-time-collaboration-<version>/
+		 * and check what versions are available.
+		 */
+		$versions = $this->get_versions();
```
We should consider pulling this out into a utility, as the RDB plugin uses this as well.

ingeniumed commented on integrations/real-time-collaboration.php:
```diff
@@ -0,0 +1,111 @@
+		$env_config = $this->get_env_config();
+
+		// Set up WebSocket authentication secret constant
+		if ( isset( $env_config['web_socket_auth_secret'] ) && ! defined( 'VIP_RTC_WS_AUTH_SECRET' ) ) {
```
I'm wondering if we should add in a check to make sure our plugin doesn't load if these aren't defined? I know we have a check already in our code, but this might be a good guard to have in place as well.

shekharnwagh replied:
Pushed a commit to add `can_load` function which does such checks.

shekharnwagh committed:
Guard against multiple instances of Gutenberg plugin

Activate the real-time collaboration integration when the plugins are
loaded and Gutenberg is not active as a plugin. It guards against the
conflicts that can arise from the custom build of Gutenberg being loaded
(required for the integration to work) and the Gutenberg plugin being
active as a plugin.

shekharnwagh committed:
Add WebSocket config validation before loading integration

Adds checks to ensure the VIP_RTC_WS_AUTH_SECRET and VIP_RTC_WS_URL
constants are defined before loading the integration. Refactors the
load() function to extract out logic for validation and constructing
the path for Gutenberg and the RTC plugin.

ingeniumed commented on integrations/real-time-collaboration.php:
```diff
@@ -0,0 +1,146 @@
+			define( 'VIP_RTC_WS_AUTH_SECRET', $env_config['web_socket_auth_secret'] );
+		}
+
+		// Set up WebSocket URL constant with dummy value
```
Minor: is it a dummy value or the actual value?

shekharnwagh replied:
Its actual value. It was using a dummy value as when the PR was opened the config didn't pass the WebSocket URL through. Fixed the comment.

shekharnwagh replied:
Thank you for catching this !

shekharnwagh committed:
fix comments around placeholder values

shekharnwagh committed:
Merge branch 'develop' into add/vip-1715-real-time-collaboration

shekharnwagh committed:
Merge branch 'develop' into add/vip-1715-real-time-collaboration
